pkg/testing/integration/program.go

```diff
@@ -1821,40 +1821,65 @@ func (pt *ProgramTester) copyTestToTemporaryDirectory() (string, string, error)
 
 	if pt.opts.LocalProviders != nil {
 		for _, provider := range pt.opts.LocalProviders {
+			// LocalProviders are relative to the working directory when running tests, NOT relative to the
+			// Pulumi.yaml. This is a bit odd, but makes it easier to construct the required paths in each
+			// test.
+			absPath, err := filepath.Abs(provider.Path)
+			if err != nil {
+				return "", "", fmt.Errorf("could not get absolute path for plugin %s: %w", provider.Path, err)
+			}
+
 			projinfo.Proj.Plugins.Providers = append(projinfo.Proj.Plugins.Providers, workspace.PluginOptions{
 				Name: provider.Package,
-				Path: provider.Path,
+				Path: absPath,
 			})
 		}
 	}
 
+	// Return a fixed up path if it's relative to sourceDir but not beneath it, else just returns the input
+	fixupPath := func(path string) (string, error) {
+		if filepath.IsAbs(path) {
+			return path, nil
+		}
+		absSource, err := filepath.Abs(sourceDir)
+		if err != nil {
+			return "", err
+		}
+		absPlugin, err := filepath.Abs(filepath.Join(absSource, path))
+		if err != nil {
+			return "", err
+		}
+		rel, err := filepath.Rel(absSource, absPlugin)
+		if err != nil {
+			return "", err
+		}
+		if strings.HasPrefix(rel, "..") {
+			return absPlugin, nil
+		}
+		return path, nil
+	}
+
 	if projinfo.Proj.Plugins != nil {
 		for i, provider := range projinfo.Proj.Plugins.Providers {
-			if !filepath.IsAbs(provider.Path) {
-				path, err := filepath.Abs(provider.Path)
-				if err != nil {
-					return "", "", fmt.Errorf("could not get absolute path for plugin %s: %w", provider.Path, err)
-				}
-				projinfo.Proj.Plugins.Providers[i].Path = path
+			path, err := fixupPath(provider.Path)
+			if err != nil {
+				return "", "", fmt.Errorf("could not get fixed path for plugin %s: %w", provider.Path, err)
```

Comment:
optional cleanup: the bodies of these three loops and the types they operate on are the same.
we could probably do:

```go
optionSets := [][]workspace.PluginOptions{
  projinfo.Proj.Plugins.Providers
  projinfo.Proj.Plugins.Languages,
  projinfo.Proj.Plugins.Analyzers,
}
for _, options := range optionSets {
  for i, opt := range options {
    path, err := fixupPath(opt.Path)
    // ...
    options[i].Path = path
  }
}
```

---

also optional: if typing projinfo.Proj.Plugins starts to get annoying, we can use the `if` here:

```go
if plugs := projinfo.Proj.Plugins; plugs != nil {
  optionSets := [][]workspace.PluginOptions{
    plugs.Providers,
    plugs.Languages,
    plugs.Analyzers,
  }
}
```